ext/drmgr/drmgr.c

```diff
@@ -3253,6 +3272,24 @@ drmgr_get_emulated_instr_data(instr_t *instr, emulated_instr_t *emulated)
     return true;
 }
 
+DR_EXPORT
+bool
+drmgr_in_emulation_region(void *drcontext, OUT emulated_instr_t *emulation_info)
+{
+    per_thread_t *pt = (per_thread_t *)drmgr_get_tls_field(drcontext, our_tls_idx);
+    if (drmgr_current_bb_phase(drcontext) != DRMGR_PHASE_INSERTION)
+        return false;
+    if (!pt->in_emulation_region)
+        return false;
+    if (emulation_info != NULL) {
+        size_t size = emulation_info->size;
+        if (size > pt->emulation_info.size)
+            size = pt->emulation_info.size;
+        memcpy(emulation_info, &pt->emulation_info, size);
```

Comment:
The memcpy is only executed when actually inside an emulation region which would normally be quite rare.  Right?

Reply:
Well, the emulation sequence for `vpscatterdd {%k1} %xmm10 -> 0x00402022(,%xmm11,4)[4byte] %k1`, which consists of 4 scalar writes consists of ~200 instrs. With a zmm operand, it's 16 scalar writes, so maybe ~800 instrs.

So, each scatter/gather could perhaps be expensive, even if rare.

Reply:
I guess we don't need to invest time in optimising this right away. But maybe worth an XXX comment?

Reply:
Woah: hundreds of instrs is going to break some things.  How many are app instrs?  Some clients insert clean calls for every app instr.  Hundreds won't fit in one block.  16, plus another 800: I guess 16 clean calls will fit?  Not sure how many bytes per clean call.  Maybe worth measuring.

Our internal use sets `-max_bb_instrs 256` which at first I thought would cause problems with this expansion: but that's enforced by DR prior to the app2app phase.  So I guess that will work.

Back to this PR:
+ The structure is 4*8=32 bytes: so not huge.  (Actually: I see a maybe corner-case problem in the code: the memcpy should not clobber the input's size field as that could cause problems if there were some iteration -- though there shouldn't be as the callback is called just once; but maybe some client would invoke `drmgr_in_emulation_region` twice in the same callback.  So it should be 24 bytes copied.)
+ We need to look at labels: but we could conceivably ignore non-app non-label instrs, which would skip how many of the expansion instrs for scatter?
+ Maybe there doesn't need to be a memcpy -- should the interface return a *pointer* to drmgr's copy of the struct?

Reply:
Yeah actually I think returning a pointer to drmgr's copy should be fine.  I'm going to make that change: better to change the interface now.

Reply:
Here's  a gist of the expansion for `vpscatterdd {%k1} %xmm10 -> 0x00402022(,%xmm11,4)[4byte] %k1`: https://gist.github.com/abhinav92003/4a78ec4ad1621cbf6ca905a6774cb389.

I see ~200 instrs (out of which 41 are app) and 131 labels.

Reply:
I think I should clarify: the expansion above is after drcachesim instrumentation too. (Apologies if that caused confusion about size of the scatter/gather expansion).

Reply:
Here's the expansion of the same instr without any other instrumentation: https://gist.github.com/abhinav92003/198dc8ab3933eacfa3b4703ca8f81363

70 instrs (45 app) and 20 labels.

Reply:
OK that sounds more like what I would expect: hundreds after drcachesim, but "just" dozens in the original expansion.

Reply:
All right now there's no memcpy: the output param is a const**.